### ISSUE TITLE
Cannot Invoke Chrome browser inside Jenkins. 

### ISSUE THREAD

**Issue by DinakarC-practise** [GoogleChrome/lighthouse#10088]
Running Lighthouse programmatically. It runs well on my local, when trying to invoke this is inside Jenkins. I see the error for chrome launcher. 

#### What is the current behavior?
Running Audit ( 1 of 2 ) for https://www.google.com/
(node:35) UnhandledPromiseRejectionWarning: Error
    at new LauncherError (/xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/utils.js:35:22)
    at new ChromePathNotSetError (/xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/utils.js:42:9)
    at Object.linux (/xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/chrome-finder.js:108:15)
    at Function.getInstallations (/xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/chrome-launcher.js:114:51)
    at Launcher.<anonymous> (/xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/chrome-launcher.js:147:48)
    at Generator.next (<anonymous>)
    at /xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/chrome-launcher.js:12:71
    at new Promise (<anonymous>)
    at __awaiter (/xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/chrome-launcher.js:8:12)
    at Launcher.launch (/xxxx/jenkins/workspace/-fetching-metrics-for-the-builds/node_modules/chrome-launcher/dist/chrome-launcher.js:135:16)
(node:35) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:35) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

My JS file invoking lighthouse. 
`const lighthouse = require('lighthouse');
const chromeLauncher = require('chrome-launcher');
const configSetting = require('./custom-config.js');
const writeJsonFile = require('write-json-file');
const fs = require('fs');
function launchChromeAndRunLighthouse(url, opts, config = null) {
  return chromeLauncher.launch({chromeFlags: opts.chromeFlags}).then(chrome => {
    opts.port = chrome.port;
    return lighthouse(url, opts, config).then(results => {
      return chrome.kill().then(() => results.lhr)
    });
  });
}
const opts = {
    chromeFlags: ['--headless'],
    // chromeFlags: ['--no-sandbox','--headless'],
	onlyCategories: ['performance', 'pwa']
};`

My package .json. 
`{
    "dependencies": {
        "chrome": "^0.1.0",
        "chrome-launcher": "^0.12.0",
        "lighthouse": "^5.6.0",
        "write-json-file": "^4.2.1"
    }
}`
Jenkins file
`	steps{
				sh "npm install"
			}
        }
        stage('run node'){
            steps{
                sh "node index.js"
            }
        }`
#### Environment Information
* Affected Channels: 
* Lighthouse version: 5.6
* Chrome-launcher version:0.12 
* Node.js version: 10.17
* Operating System: Jenkins. 

**Related issues**

**Comment by patrickhulce** [comment]
It's throwing a `ChromePathNotSetError` which means you don't have Chrome installed in the expected path on your Jenkins machine or you haven't set `CHROME_PATH` variable that tells Lighthouse where Chrome is.

**Comment by DinakarC-practise** [comment]
Do I have to point to the chrome-launcher in node modules for chrome path or do I have to install separately?

**Comment by patrickhulce** [comment]
Chrome is not installed through node_modules or via `chrome-launcher` you'll need to do that through your OS's package manager. I'm assuming `yum` or `apt-get` if this is a jenkins box on a linux OS.

**Comment by DinakarC-practise** [comment]
Hey @patrickhulce thanks for the advice and I tried installing chrome using chrome driver. Here is the updated code I used. `const lighthouse = require('lighthouse');
const chromeLauncher = require('chrome-launcher');
const configSetting = require('./custom-config.js');
const writeJsonFile = require('write-json-file');
var childProcess = require('child_process');
var chromedriver = require('chromedriver');
var binPath = chromedriver.path;
const fs = require('fs');
let config = null;
function launchChromeAndRunLighthouse(url, opts, config) {
  return chromeLauncher.launch({chromeFlags: opts.chromeFlags, chromePath: binPath}).then(chrome => {
    opts.port = chromePort;
    return lighthouse(url, opts, config).then(results => {
      return chrome.kill().then(() => results.lhr)
    });
  });
}
const opts = {
	chromeFlags: ['--headless'],
    // chromeFlags: ['--no-sandbox','--headless'],
	onlyCategories: ['performance', 'pwa']
};` but still showing the error `(node:49) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:33649
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1107:14)
(node:49) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:49) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

I tried giving different port number there too.

**Comment by patrickhulce** [comment]
It'd hard to figure out which options you're actually trying to use @DinakarC-practise. Also take a look at the Chrome output to see what it's complaining about.

```
chromeLauncher.launch({
  chromePath: binPath,
  chromeFlags: ['--headless', '--disable-gpu', '--no-sandbox'],
  userDataDir: '/tmp/chrome-1',
})
```

Now `/tmp/chrome-1/chrome-out.log` and `/tmp/chrome-1/chrome-err.log` should contain why Chrome failed to start.

**Comment by DinakarC-practise** [comment]
Hey @patrickhulce this is what I see inside chrome-out.log ````
Starting ChromeDriver 79.0.3945.36 (3582db32b33893869b8c1339e8f4d9ed1816f143-refs/branch-heads/3945@{#614}) on port 9515
Only local connections are allowed.
Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.
```

**Comment by patrickhulce** [comment]
@DinakarC-practise you're passing the executable path to ChromeDriver, not Chrome. Lighthouse and chrome-launcher need to talk directly to Chrome not through ChromeDriver. 

Reading back through your previous comments it sounds like you still haven't installed Chrome, just the `chromedriver` npm module?

**Comment by DinakarC-practise** [comment]
I have installed chrome_launcher and I see them in my node modules too. 
` "dependencies": {
        "chrome": "^0.1.0",
        "chrome-launcher": "^0.12.0",
        "chromedriver": "^79.0.0",
        "chromium": "^2.1.1",
        "lighthouse": "^5.6.0",
        "write-json-file": "^4.2.1"
    }`

**Comment by DinakarC-practise** [comment]
Is there any other way to communicate with you I cannot really post my source code here coz that will be violating rules.
And FYI I am having a conversation with `gmimani@google.com` from google for all these issues and also I requested her for a session with one of the lighthouse dev, I don't know the status of that request but It will be helpful if you and me can have session on this issue and I can explain our use case on lighthouse.

**Comment by DinakarC-practise** [comment]
Hey @patrickhulce I tried running chrome using dockers and health checks are passed for chrome installation, but the error says chrome launcher waiting for browser. 
`+ npx lhci autorun --collect.url=https://nodejs.org/
✅  .lighthouseci/ directory writable
⚠️   Configuration file not found
✅  Chrome installation found
Healthcheck passed!
Running Lighthouse 3 time(s) on https://nodejs.org/
Run #1...failed!
Error: Lighthouse failed with exit code 1
    at ChildProcess.child.on.code (/usr/local/lib/node_modules/@lhci/cli/src/collect/lighthouse-runner.js:103:21)
    at ChildProcess.emit (events.js:197:13)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:254:12)
Fri, 20 Dec 2019 18:59:15 GMT ChromeLauncher Waiting for browser.
Fri, 20 Dec 2019 18:59:15 GMT ChromeLauncher Waiting for browser...
Fri, 20 Dec 2019 18:59:15 GMT ChromeLauncher Waiting for browser.....
Fri, 20 Dec 2019 18:59:16 GMT ChromeLauncher Waiting for browser.......
Fri, 20 Dec 2019 18:59:16 GMT ChromeLauncher Waiting for browser.........`

My docker file for installing chrome, lighthouse
 `RUN apt-get update && apt-get install -y \
  apt-transport-https \
  ca-certificates \
  curl \
  gnupg \
  --no-install-recommends \
  && curl -sSL https://deb.nodesource.com/setup_12.x | bash - \
  && curl -sSL https://dl.google.com/linux/linux_signing_key.pub | apt-key add - \
  && echo "deb https://dl.google.com/linux/chrome/deb/ stable main" > /etc/apt/sources.list.d/google-chrome.list \
  && apt-get update && apt-get install -y \
  google-chrome-stable \
  fontconfig \
  fonts-ipafont-gothic \
  fonts-wqy-zenhei \
  fonts-thai-tlwg \
  fonts-kacst \
  fonts-symbola \
  fonts-noto \
  fonts-freefont-ttf \
  nodejs \
  --no-install-recommends \
  && apt-get purge --auto-remove -y curl gnupg \
  && rm -rf /var/lib/apt/lists/*
ARG CACHEBUST=1
# Installing lighthouse latest version
RUN npm install -g lighthouse
# Installing LHCI latest version
RUN npm install -g @lhci/cli`

**Comment by patrickhulce** [comment]
Discussed offline, this amounts to a duplicate of https://github.com/GoogleChrome/lighthouse-ci/issues/102#issuecomment-555264580 and is about installing Chrome.

tl;dr use an already working docker image and the `--no-sandbox` Chrome flag or wait for https://github.com/GoogleChrome/lighthouse-ci/issues/164 to be completed.
